addon/internal-session.js

```diff
@@ -44,59 +45,87 @@ export default Ember.ObjectProxy.extend(Ember.Evented, {
-          this.set('content', restoredContent);
-          this._clear();
-          reject();
-        });
-      } else {
-        delete (restoredContent || {}).authenticated;
-        this.set('content', restoredContent);
-        this._clear();
-        reject();
-      }
+          this._clear().then(reject, reject);
+        }
+      }, () => {
+        this._clear().then(reject, reject);
+      });
     });
   },
 
+  _callStoreAsync(method, ...params) {
+    let result = this.store[method].apply(this.store, params);
+    if (typeof result === 'undefined' || typeof result.then === 'undefined') {
+      Ember.deprecate(`Ember Simple Auth: Synchronous stores have been deprecated. Make sure your custom store's ${method} method returns a promise.`, false, {
+        id: `ember-simple-auth.session-store.synchronous-${method}`,
+        until: '2.0.0'
+      });
+      return RSVP.Promise.resolve(result);
+    } else {
+      return result;
+    }
+  },
+
   _setup(authenticator, authenticatedContent, trigger) {
-    trigger = !!trigger && !this.get('isAuthenticated');
-    this.beginPropertyChanges();
-    this.setProperties({
-      isAuthenticated: true,
-      authenticator
+    return new RSVP.Promise((resolve, reject) => {
+      trigger = !!trigger && !this.get('isAuthenticated');
+      this.beginPropertyChanges();
+      this.setProperties({
+        isAuthenticated: true,
+        authenticator
+      });
+      Ember.set(this.content, 'authenticated', authenticatedContent);
+      this._bindToAuthenticatorEvents();
+      this._updateStore().then(() => {
+        this.endPropertyChanges();
+        if (trigger) {
+          this.trigger('authenticationSucceeded');
+        }
+        resolve();
+      }, () => {
+        this.setProperties({
+          isAuthenticated: false,
+          authenticator: null
+        });
+        Ember.set(this.content, 'authenticated', {});
+        reject();
+      });
     });
-    Ember.set(this.content, 'authenticated', authenticatedContent);
-    this._bindToAuthenticatorEvents();
-    this._updateStore();
-    this.endPropertyChanges();
-    if (trigger) {
-      this.trigger('authenticationSucceeded');
-    }
   },
 
   _clear(trigger) {
-    trigger = !!trigger && this.get('isAuthenticated');
-    this.beginPropertyChanges();
-    this.setProperties({
-      isAuthenticated: false,
-      authenticator:   null
+    return new RSVP.Promise((resolve, reject) => {
+      trigger = !!trigger && this.get('isAuthenticated');
+      this.beginPropertyChanges();
+      this.setProperties({
+        isAuthenticated: false,
+        authenticator:   null
+      });
+      Ember.set(this.content, 'authenticated', {});
+      this._updateStore().then(() => {
+        this.endPropertyChanges();
+        if (trigger) {
+          this.trigger('invalidationSucceeded');
+        }
+        resolve();
+      }, reject);
```

Comment:
See above - I think there must be a `this.endPropertyChanges()` in the reject case as well.